tests/async/test_browsercontext_add_cookies.py

```diff
@@ -0,0 +1,387 @@
+
+    server.set_route("/empty.html", handler)
+    await context.addCookies(
+        [{"url": server.EMPTY_PAGE, "name": "cookie", "value": "value"}]
+    )
+    page = await context.newPage()
+    await page.goto(server.EMPTY_PAGE)
+    assert cookie == ["cookie=value"]
+
+
+async def test_should_isolate_cookies_in_browser_contexts(context, server, browser):
+    another_context = await browser.newContext()
+    await context.addCookies(
+        [{"url": server.EMPTY_PAGE, "name": "isolatecookie", "value": "page1value"}]
+    )
+    await another_context.addCookies(
+        [{"url": server.EMPTY_PAGE, "name": "isolatecookie", "value": "page2value"}]
+    )
+
+    cookies_1 = await context.cookies()
+    cookies_2 = await another_context.cookies()
+    assert len(cookies_1) == 1
+    assert len(cookies_2) == 1
+    assert cookies_1[0]["name"] == "isolatecookie"
+    assert cookies_1[0]["value"] == "page1value"
+    assert cookies_2[0]["name"] == "isolatecookie"
+    assert cookies_2[0]["value"] == "page2value"
+    await another_context.close()
+
+
+async def test_should_isolate_session_cookies(context, server, browser):
+    server.set_route(
+        "/setcookie.html",
+        lambda r: (r.setHeader("Set-Cookie", "session=value"), r.finish(),),
+    )
+
+    page_1 = await context.newPage()
+    await page_1.goto(server.PREFIX + "/setcookie.html")
+    ##
+    page_2 = await context.newPage()
+    await page_2.goto(server.EMPTY_PAGE)
+    cookies_2 = await context.cookies()
+    assert len(cookies_2) == 1
+    assert ",".join(list(map(lambda c: c["value"], cookies_2))) == "value"
+    ##
+    context_b = await browser.newContext()
+    page_3 = await context_b.newPage()
+    await page_3.goto(server.EMPTY_PAGE)
+    cookies_3 = await context_b.cookies()
+    assert cookies_3 == []
+    await context_b.close()
+
+
+async def test_should_isolate_persistent_cookies(context, server, browser):
+    server.set_route(
+        "/setcookie.html",
+        lambda r: (
+            r.setHeader("Set-Cookie", "persistent=persistent-value; max-age=3600"),
+            r.finish(),
+        ),
+    )
+
+    page = await context.newPage()
+    await page.goto(server.PREFIX + "/setcookie.html")
+
+    context_1 = context
+    context_2 = await browser.newContext()
+    [page_1, page_2] = await asyncio.gather(context_1.newPage(), context_2.newPage())
+    await asyncio.gather(page_1.goto(server.EMPTY_PAGE), page_2.goto(server.EMPTY_PAGE))
+    [cookies_1, cookies_2] = await asyncio.gather(
+        context_1.cookies(), context_2.cookies()
+    )
+    assert len(cookies_1) == 1
+    assert cookies_1[0]["name"] == "persistent"
+    assert cookies_1[0]["value"] == "persistent-value"
+    assert len(cookies_2) == 0
+    await context_2.close()
+
+
+async def test_should_isolate_send_cookie_header(server, context, browser):
+    cookie = []
+
+    def handler(request):
+        cookie.extend(request.requestHeaders.getRawHeaders("cookie") or [])
```

Comment:
In the JS version, cookie was simply assigned to. In Python, that didn't work so I used a mutable list.